File: intrinsic/assets/data/fake_data_assets.go
// Copyright 2023 Intrinsic Innovation LLC

// Package fakedataassets provides a fake DataAssets service.
package fakedataassets

import (
	"context"
	"slices"
	"strings"
	"testing"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/local"
	"google.golang.org/grpc/status"
	"intrinsic/assets/idutils"
	"intrinsic/testing/grpctest"

	dapb "intrinsic/assets/data/proto/v1/data_asset_go_proto"
	dasgrpcpb "intrinsic/assets/data/proto/v1/data_assets_go_grpc_proto"
	daspb "intrinsic/assets/data/proto/v1/data_assets_go_grpc_proto"
	mpb "intrinsic/assets/proto/metadata_go_proto"
)

const defaultPageSize = 20

// Service is a fake implementation of the DataAssets service.
type Service struct {
	dataAssets map[string]*dapb.DataAsset
}

func (s *Service) ListDataAssets(ctx context.Context, req *daspb.ListDataAssetsRequest) (*daspb.ListDataAssetsResponse, error) {
	var filteredAssets []*dapb.DataAsset
	for _, da := range s.dataAssets {
		if req.GetStrictFilter().GetProtoName() != "" {
			typeURL := da.GetData().GetTypeUrl()
			if !strings.HasSuffix(typeURL, req.GetStrictFilter().GetProtoName()) {
				continue
			}
		}
		filteredAssets = append(filteredAssets, da)
	}

	// Sort by ID for consistent pagination.
	slices.SortFunc(filteredAssets, func(a, b *dapb.DataAsset) int {
		idA := idutils.IDFromProtoUnchecked(a.GetMetadata().GetIdVersion().GetId())
		idB := idutils.IDFromProtoUnchecked(b.GetMetadata().GetIdVersion().GetId())
		return strings.Compare(idA, idB)
	})

	// Determine the start of the page.
	offset := 0
	if req.GetPageToken() != "" {
		assetFound := false
		for i, da := range filteredAssets {
			id := idutils.IDFromProtoUnchecked(da.GetMetadata().GetIdVersion().GetId())
			if id == req.GetPageToken() {
				assetFound = true
				offset = i
				break
			}
		}
		if !assetFound {
			return nil, status.Errorf(codes.InvalidArgument, "invalid page token: %q", req.GetPageToken())
		}
	}

	pageSize := defaultPageSize
	if req.GetPageSize() > 0 {
		pageSize = int(req.GetPageSize())
	}
	lastIndex := min(offset+pageSize, len(filteredAssets)) - 1

	nextPageToken := ""
	if lastIndex < len(filteredAssets)-1 {
		nextPageToken = idutils.IDFromProtoUnchecked(filteredAssets[lastIndex+1].GetMetadata().GetIdVersion().GetId())
	}

	return &daspb.ListDataAssetsResponse{
		DataAssets:    filteredAssets[offset : lastIndex+1],
		NextPageToken: nextPageToken,
	}, nil
}

func (s *Service) ListDataAssetMetadata(ctx context.Context, req *daspb.ListDataAssetMetadataRequest) (*daspb.ListDataAssetMetadataResponse, error) {
	listRequest := &daspb.ListDataAssetsRequest{
		StrictFilter: req.GetStrictFilter(),
		PageSize:     req.GetPageSize(),
		PageToken:    req.GetPageToken(),
	}
	listResponse, err := s.ListDataAssets(ctx, listRequest)
	if err != nil {
		return nil, err
	}

	metadata := make([]*mpb.Metadata, len(listResponse.GetDataAssets()))
	for i, da := range listResponse.GetDataAssets() {
		metadata[i] = da.GetMetadata()
	}
	return &daspb.ListDataAssetMetadataResponse{
		Metadata:      metadata,
		NextPageToken: listResponse.GetNextPageToken(),
	}, nil
}

func (s *Service) GetDataAsset(ctx context.Context, req *daspb.GetDataAssetRequest) (*dapb.DataAsset, error) {
	id := idutils.IDFromProtoUnchecked(req.GetId())
	da, ok := s.dataAssets[id]
	if !ok {
		return nil, status.Errorf(codes.NotFound, "data asset not found: %q", id)
	}
	return da, nil
}

func (s *Service) StreamReferencedData(req *daspb.StreamReferencedDataRequest, stream dasgrpcpb.DataAssets_StreamReferencedDataServer) error {
	return status.Errorf(codes.Unimplemented, "not implemented")
}

// Fake provides a fake DataAssets service.
type Fake struct {
	Client  dasgrpcpb.DataAssetsClient
	Service *Service
	Server  *grpc.Server
}

type startServerOpts struct {
	dataAssets []*dapb.DataAsset
}

// StartServerOpt is a functional option for StartServer.
type StartServerOpt func(*startServerOpts)

// WithDataAssets returns a StartServerOpt that sets the data assets.
func WithDataAssets(dataAssets []*dapb.DataAsset) StartServerOpt {
	return func(opts *startServerOpts) {
		opts.dataAssets = dataAssets
	}
}

// StartServer creates a gRPC server with the DataAssets service registered.
func StartServer(ctx context.Context, t *testing.T, options ...StartServerOpt) *Fake {
	t.Helper()

	opts := startServerOpts{}
	for _, opt := range options {
		opt(&opts)
	}

	dataAssetsMap := make(map[string]*dapb.DataAsset)
	for _, da := range opts.dataAssets {
		id := idutils.IDFromProtoUnchecked(da.GetMetadata().GetIdVersion().GetId())
		if _, ok := dataAssetsMap[id]; ok {
			t.Fatalf("Duplicate data asset ID: %v", id)
		}
		dataAssetsMap[id] = da
	}

	server := grpc.NewServer()
	service := &Service{
		dataAssets: dataAssetsMap,
	}
	dasgrpcpb.RegisterDataAssetsServer(server, service)

	srvAddr := grpctest.StartServerT(t, server)
	conn, err := grpc.NewClient(srvAddr, grpc.WithTransportCredentials(local.NewCredentials()))
	if err != nil {
		t.Fatalf("failed to dial test server: %v", err)
	}
	t.Cleanup(func() { conn.Close() })

	return &Fake{
		Client:  dasgrpcpb.NewDataAssetsClient(conn),
		Service: service,
		Server:  server,
	}
}

File: intrinsic/assets/data/fake_data_assets.py
# Copyright 2023 Intrinsic Innovation LLC

"""Fake implementation of the DataAssets service."""

from typing import Callable
from typing import NoReturn
from typing import Sequence

import grpc
from grpc.framework.foundation import logging_pool

from intrinsic.assets import id_utils
from intrinsic.assets.data.proto.v1 import data_asset_pb2
from intrinsic.assets.data.proto.v1 import data_assets_pb2
from intrinsic.assets.data.proto.v1 import data_assets_pb2_grpc

_default_page_size = 20

CleanupFunction = Callable[[], None]


class FakeDataAssetsService(data_assets_pb2_grpc.DataAssetsServicer):
  """Fake implementation of the DataAssets service."""

  _data_assets: dict[str, data_asset_pb2.DataAsset]

  @classmethod
  def start_server(
      cls, data_assets: Sequence[data_asset_pb2.DataAsset], port: int = 0
  ) -> tuple[data_assets_pb2_grpc.DataAssetsStub, CleanupFunction]:
    """Creates a FakeDataAssetsService.

    Args:
      data_assets: The data assets to serve.
      port: The port on which to listen, or 0 to pick an unused port.

    Returns:
      A tuple of the DataAssets stub and a cleanup function to close the
      channel and stop the server.
    """
    service = FakeDataAssetsService(data_assets)
    server = grpc.server(
        logging_pool.pool(max_workers=1),
        options=[("grpc.max_receive_message_length", -1)],
    )
    data_assets_pb2_grpc.add_DataAssetsServicer_to_server(service, server)
    port = server.add_secure_port(
        f"[::]:{port}", grpc.local_server_credentials()
    )
    server.start()

    channel = grpc.secure_channel(
        f"localhost:{port}", grpc.local_channel_credentials()
    )
    stub = data_assets_pb2_grpc.DataAssetsStub(channel)

    def cleanup():
      channel.close()
      server.stop(None)

    return stub, cleanup

  def __init__(self, data_assets: Sequence[data_asset_pb2.DataAsset]):
    self._data_assets = {}

    for data_asset in data_assets:
      asset_id = id_utils.id_from_proto(data_asset.metadata.id_version.id)
      if asset_id in self._data_assets:
        raise ValueError(f"Duplicate data asset ID: {asset_id}")

      self._data_assets[asset_id] = data_asset

  def ListDataAssets(
      self,
      request: data_assets_pb2.ListDataAssetsRequest,
      context: grpc.ServicerContext,
  ) -> data_assets_pb2.ListDataAssetsResponse:
    """Lists data assets."""
    filtered_assets = []
    for data_asset in self._data_assets.values():
      if request.strict_filter is not None and request.strict_filter.proto_name:
        type_url = data_asset.data.type_url
        if type_url.split("/")[-1] != request.strict_filter.proto_name:
          continue
      filtered_assets.append(data_asset)

    # Sort by ID for consistent pagination.
    filtered_assets.sort(
        key=lambda asset: id_utils.id_from_proto(asset.metadata.id_version.id)
    )

    # Determine the start of the page.
    offset = 0
    if request.page_token:
      asset_found = False
      for i, asset in enumerate(filtered_assets):
        if (
            id_utils.id_from_proto(asset.metadata.id_version.id)
            == request.page_token
        ):
          asset_found = True
          offset = i
          break
      if not asset_found:
        _abort_with_status(
            context=context,
            code=grpc.StatusCode.INVALID_ARGUMENT,
            message=f"Invalid page token: {request.page_token}",
        )

    page_size = (
        request.page_size if request.page_size > 0 else _default_page_size
    )
    last_index = min(offset + page_size, len(filtered_assets)) - 1

    next_page_token = ""
    if last_index < len(filtered_assets) - 1:
      next_page_token = id_utils.id_from_proto(
          filtered_assets[last_index + 1].metadata.id_version.id
      )

    return data_assets_pb2.ListDataAssetsResponse(
        data_assets=filtered_assets[offset : last_index + 1],
        next_page_token=next_page_token,
    )

  def ListDataAssetMetadata(
      self,
      request: data_assets_pb2.ListDataAssetMetadataRequest,
      context: grpc.ServicerContext,
  ) -> data_assets_pb2.ListDataAssetMetadataResponse:
    """Lists data asset metadata."""
    list_request = data_assets_pb2.ListDataAssetsRequest(
        strict_filter=request.strict_filter,
        page_size=request.page_size,
        page_token=request.page_token,
    )
    list_response = self.ListDataAssets(list_request, context)

    return data_assets_pb2.ListDataAssetMetadataResponse(
        metadata=[asset.metadata for asset in list_response.data_assets],
        next_page_token=list_response.next_page_token,
    )

  def GetDataAsset(
      self,
      request: data_assets_pb2.GetDataAssetRequest,
      context: grpc.ServicerContext,
  ) -> data_asset_pb2.DataAsset:
    """Gets a data asset."""
    try:
      return self._data_assets[id_utils.id_from_proto(request.id)]
    except KeyError as e:
      _abort_with_status(
          context=context,
          code=grpc.StatusCode.NOT_FOUND,
          message=f"Data asset not found: {e}",
      )

  def StreamReferencedData(
      self,
      request: data_assets_pb2.StreamReferencedDataRequest,
      context: grpc.ServicerContext,
  ) -> data_assets_pb2.StreamReferencedDataResponse:
    """Streams referenced data."""
    del request  # Unused.
    _abort_with_status(
        context=context,
        code=grpc.StatusCode.UNIMPLEMENTED,
        message=(
            "StreamReferencedData is not implemented in FakeDataAssetsService."
        ),
    )


def _abort_with_status(
    context: grpc.ServicerContext,
    code: grpc.StatusCode,
    message: str,
) -> NoReturn:
  context.abort(code, message)

  # This will never be raised, but we need it to satisfy static type checking,
  raise AssertionError("This error should not have been raised.")
